### ISSUE TITLE
struct members sometimes refer to wrong struct types

### ISSUE THREAD

**Issue by wseaton** [kube-rs/kopium#28]
Tried to run kopium on: 

```sh
kopium applications.argoproj.io --docs > argoapp.rs
```

Get the following due to some of the CRD's being self-referential:

```
error[E0072]: recursive type `ApplicationStatusSyncComparedToSourceDirectory` has infinite size
    --> src/argoapp.rs:1330:1
     |
1330 | pub struct ApplicationStatusSyncComparedToSourceDirectory {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ recursive type has infinite size
...
1336 |     pub jsonnet: Option<ApplicationStatusSyncComparedToSourceDirectory>,
     |                  ------------------------------------------------------ recursive without indirection
     |
help: insert some indirection (e.g., a `Box`, `Rc`, or `&`) to make `ApplicationStatusSyncComparedToSourceDirectory` representable
     |
1336 |     pub jsonnet: Box<Option<ApplicationStatusSyncComparedToSourceDirectory>>,
     |                  ++++                                                      +
```

I think I can fixup these myself by `Box`ing them, any other ideas?

**Comment by wseaton** [comment]
Hmm, something seems more off, looking more carefully at: 

```rust
/// Operation contains information about a requested or running operation
#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ApplicationOperation {
    /// Info is a list of informational items for this operation
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub info: Vec<ApplicationOperationInfo>,
    /// InitiatedBy contains information about who initiated the operations
    pub initiatedBy: Option<ApplicationOperation>,
    /// Retry controls the strategy to apply if a sync fails
    pub retry: Option<ApplicationOperation>,
    /// Sync contains parameters for the operation
    pub sync: Option<ApplicationOperation>,
}
```

Some of the intermediate types don't look correct when compared to the upstream CRD here: https://github.com/argoproj/argo-cd/blob/master/manifests/crds/application-crd.yaml#L48-L77

**Comment by clux** [comment]
hm, yeah, some of those do look weird. looks like there's a mistake in the logic for combining suffixes from names.

it has created the intermediate structs:

```rust

#[derive(Serialize, Deserialize, Clone, Debug)]
pub struct ApplicationOperationInitiatedBy {
    pub automated: Option<bool>,
    pub username: Option<String>,
}
```

but referenced the wrong ones (without `InitedBy` suffix)

also the self-referential thing is probably also something we need to do better - if Boxing works then that's probably the right thing to do.

thanks for the report!

**Comment by clux** [comment]
working on a pr in #29 - it should chain on the names correctly - but it's not done yet.
probably going to put argo app as a test case, just need an example application instance to put in the test cluster

**Comment by wseaton** [comment]
@clux I manually did a fixup of that CRD and it works fine w/ `kube-rs` (there was actually no recursion), here's the diff if you want to use it as a test case: https://gist.github.com/wseaton/426cabf4915d429f8e7f0b6119bd846e

**Comment by clux** [comment]
Appreciate it. I have added tests for argocd application and with the simple one-line fix it seems to be sufficient to generate working structs.

**Comment by clux** [comment]
fix released in [`kopium@0.6.0`](https://github.com/kube-rs/kopium/releases/tag/v0.6.0).
